lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp

```diff
@@ -1559,6 +1559,99 @@ class BroadcastOpConversion
+
+    if (resultVectorSize != 512) {
+      op.emitWarning() << "aievec.max conversion with " << resultVectorSize
+                       << "-bit result is not supported.\n";
+      return failure();
+    }
+
+    // create xllvm intrinsic
+    Value maxOp = nullptr;
+    if (llvm::isa<IntegerType>(resultScaTy)) {
+      // create constant for cmp
+      auto cmpCst = rewriter.create<LLVM::ConstantOp>(
+          loc, rewriter.getI32Type(), rewriter.getI32IntegerAttr(0));
+      SmallVector<Value> operands{adaptor.getLhs(), adaptor.getRhs(), cmpCst};
+      if (resultBitWidth == 8) {
+        maxOp = rewriter.create<xllvm::VectorMaxLt8IntrOp>(
+            loc,
+            mlir::LLVM::LLVMStructType::getLiteral(
+                rewriter.getContext(),
+                {VectorType::get({64}, rewriter.getI8Type()),
+                 rewriter.getI32Type()}),
+            forceCastOperandsToSignature(
+                rewriter, loc, operands,
+                {VectorType::get({64}, rewriter.getI8Type()),
+                 VectorType::get({64}, rewriter.getI8Type()),
+                 rewriter.getI32Type()}));
+      } else if (resultBitWidth == 16) {
+        maxOp = rewriter.create<xllvm::VectorMaxLt16IntrOp>(
+            loc,
+            mlir::LLVM::LLVMStructType::getLiteral(
+                rewriter.getContext(),
+                {VectorType::get({32}, rewriter.getI16Type()),
+                 rewriter.getI32Type()}),
+            forceCastOperandsToSignature(
+                rewriter, loc, operands,
+                {VectorType::get({32}, rewriter.getI16Type()),
+                 VectorType::get({32}, rewriter.getI16Type()),
+                 rewriter.getI32Type()}));
+      } else if (resultBitWidth == 32) {
+        maxOp = rewriter.create<xllvm::VectorMaxLt32IntrOp>(
+            loc,
+            mlir::LLVM::LLVMStructType::getLiteral(
+                rewriter.getContext(),
+                {VectorType::get({16}, rewriter.getI32Type()),
+                 rewriter.getI32Type()}),
+            forceCastOperandsToSignature(
+                rewriter, loc, operands,
+                {VectorType::get({16}, rewriter.getI32Type()),
+                 VectorType::get({16}, rewriter.getI32Type()),
+                 rewriter.getI32Type()}));
+      }
+    } else {
+      if (resultBitWidth == 16) {
+        maxOp = rewriter.create<xllvm::VectorMaxLtBf16IntrOp>(
+            loc,
+            mlir::LLVM::LLVMStructType::getLiteral(
+                rewriter.getContext(),
+                {VectorType::get({32}, rewriter.getBF16Type()),
+                 rewriter.getI32Type()}),
+            forceCastOperandsToSignature(
+                rewriter, loc, {adaptor.getLhs(), adaptor.getRhs()},
+                {VectorType::get({32}, rewriter.getBF16Type()),
+                 VectorType::get({32}, rewriter.getBF16Type())}));
+      }
+    }
+
+    if (!maxOp) {
+      op.emitWarning() << "aievec.max conversion is not supported.\n";
```

Comment:
The example is below, and the error message looks fine to me. It errors out when failing to legalize aievec.max. I can change it to emit an error, though, to make it a step closer to the root cause.
```
within split at /proj/xsjhdstaff1/linjames/mlir-aie/test/Conversion/AIEVecToLLVM/test-max.mlir:49 offset :17:8: warning: aievec.max conversion is not supported.

  %0 = aievec.max %arg0, %arg0 : vector<128xi4>
       ^
within split at /proj/xsjhdstaff1/linjames/mlir-aie/test/Conversion/AIEVecToLLVM/test-max.mlir:49 offset :17:8: note: see current operation: %1 = aievec.max %arg0, %arg0 : vector<128xi4>
within split at /proj/xsjhdstaff1/linjames/mlir-aie/test/Conversion/AIEVecToLLVM/test-max.mlir:49 offset :17:8: error: failed to legalize operation 'aievec.max' that was explicitly marked illegal
  %0 = aievec.max %arg0, %arg0 : vector<128xi4>
       ^
within split at /proj/xsjhdstaff1/linjames/mlir-aie/test/Conversion/AIEVecToLLVM/test-max.mlir:49 offset :17:8: note: see current operation: %0 = "aievec.max"(%arg0, %arg0) : (vector<128xi4>, vector<128xi4>) -> vector<128xi4>
```

Reply:
Looks good, then 🙂At some point, we might want to give this a good thinking. This approach of "warning with info + fail to legalize error" looks good to me, but we are not consistent throughout, and there are some implications (like, what happens when a different pattern targeting the same op can actually legalize it). It's something we can decide on later; it can be easily fixed, so it doesn't bother me.

That said, "aievec.max conversion is not supported" + "failed to legalize operation 'aievec.max'" does not provide any additional information. It's a missed opportunity to explain why it can't legalize (say, "unsupported element type" or something along those lines). It might even be better to just start emitting debug information, which is probably the better approach.

This is just "food for thought", I'm happy with the original implementation 🙂

Reply:
Agree that "aievec.max conversion is not supported" isn't very helpful. I will add more debugging info on that when addressing the comment from David.